src/features/weak-references.md

```diff
@@ -54,104 +57,138 @@ A JavaScript `WeakMap` is not really _weak_: it actually refers _strongly_ to it
+**Note:** `WeakRef`s to functions must be treated with caution. JavaScript functions are [closures](https://en.wikipedia.org/wiki/Closure_(computer_programming)) and strongly reference the outer environments which contain the values of free variables referenced inside the functions. These outer environments may contain variables that _other_ closures reference as well. That is, when dealing with closures, their memory is often strongly referenced by other closures in subtle ways. Keep this in mind when writing code.
 :::
 
-Our final implementation looks like this:
+We first make the event listener and assign it to `this.listener`, so that it is strongly referenced by the `MovingAvg` instance. In other words, as long as the `MovingAvg` instance is alive, so is the event listener.
 
-```js
-const cache = new Map();
+Then, in `addWeakListener`, we create a `WeakRef` whose _target_ is the actual event listener. Inside `wrapper`, we `deref` it. Because `WeakRef`s do not prevent garbage collection of their targets if the targets do not have other strong references, we must manually dereference them to get the target. If the target has been garbage-collected in the meantime, `deref` returns `undefined`. Otherwise, the original target is returned, which is the `listener` function we then call using [optional chaining](/features/optional-chaining).
 
-const finalizationRegistry = new FinalizationRegistry((name) => {
-  const ref = cache.get(name);
-  if (ref !== undefined && ref.deref() === undefined) {
-    cache.delete(name);
-  }
-});
+Since the event listener is wrapped in a `WeakRef`, the _only_ strong reference to it is the `listener` property on the `MovingAvg` instance. That is, we've successfully tied the lifetime of the event listener to the lifetime of the `MovingAvg` instance.
+
+But there's still a problem here: we've added a level of indirection to `listener` by wrapping it a `WeakRef`, but the wrapper in `addWeakListener` is still leaking for the same reason that `listener` was leaking originally. The solution to this is the companion feature to `WeakRef`, `FinalizationRegistry`. With the new `FinalizationRegistry` API, we can register a callback to run when the garbage collector zaps a register object. Such callbacks are known as _finalizers_.
+
+:::note
+**Note:** The finalization callback does not run immediately after garbage-collecting the event listener. It either runs at some point in the future, or not at all — the spec doesn’t guarantee that it runs! Keep this in mind when writing code.
+:::
 
-function getImageCached(name) {
-  const ref = cache.get(name); // 1
-  if (ref !== undefined) { // 2
-    const deref = ref.deref();
-    if (deref !== undefined) return deref;
+We can register a callback with a `FinalizationRegistry` to remove `listenerWrapper` from the socket when the inner event listener is garbage-collected. Our final implementation looks like this:
+
+```js
+const gListenersRegistry = new FinalizationRegistry(
```

Comment:
Is it worth mentioning that gListenersRegistry here is global, because the lifetime of the registry is independent of the lifetime of the weakrefs?

Reply:
Yeah, good call. Seems worth a note.